packages/astro/src/core/render/core.ts

```diff
@@ -146,7 +144,17 @@ export async function render(
 		ssr,
 	});
 
-	let page = await renderPage(result, Component, pageProps, null);
+	let page: Awaited<ReturnType<typeof renderPage>>;
+	if (!Component.isAstroComponentFactory) {
+		const props: Record<string, any> = { ...(pageProps ?? {}), 'server:root': true };
```

Comment:
Inject `server:root` prop to signal that raw strings are OK

Reply:
The line below this calls `renderComponent` for top-level pages that aren't `.astro` or `.md`. That codepath currently strips all the `astro-fragment` elements from the final HTML unless the page is being hydrated. Top-level pages will never be hydrated, so we need to tell `renderComponent` not to do that.